### PR TITLE
[#321] kpis initialized early with landscape

### PR DESCRIPTION
**Pull request by bonndan**

### CODE REVIEW

File: src/main/java/de/bonndan/nivio/assessment/AssessmentController.java [hunk 1]
```diff
@@ -31,9 +28,7 @@ public AssessmentController(LandscapeRepository landscapeRepository, KPIFactory
             return ResponseEntity.notFound().build();
         }
 
-        Assessment assessment = new Assessment(
-                landscape.applyKPIs(factory.getConfiguredKPIs(landscape))
-        );
+        Assessment assessment = new Assessment(landscape.applyKPIs(landscape.getKpis()));
```
Comment (mfbieber) [inline src/main/java/de/bonndan/nivio/assessment/AssessmentController.java:31]: This looks a bit strange at the first sight. I checked more deeply and realized that `applyKPIs()` is a default method in the `Asseable` interface that recursively calls itself. 

I would have suggested to make `applyKPIs()` in `Landscape` use the KPIs already known to the `Landscape`. I did not see any usage of `landscape.applyKPIs(Map<String, KPI> kpis)` where we would provide the KPIs from the outside.

Comment (mfbieber) [reply src/main/java/de/bonndan/nivio/assessment/AssessmentController.java:31]: I think we could check how the `Landscape` object works after https://github.com/dedica-team/nivio/issues/212 has been resolved. Maybe thinks like this are already cleaned up after these changes.

File: src/main/java/de/bonndan/nivio/assessment/kpi/KPIFactory.java [hunk 2]
```diff
@@ -32,16 +29,18 @@ public KPIFactory(ApplicationEventPublisher eventPublisher) {
...
+     * @param kpiConfigMap the landscape kpi configurations
      * @return effective KPIs
+     * @throws ProcessingException if a KPI is misconfigured
      */
-    public Map<String, KPI> getConfiguredKPIs(Landscape landscape) {
+    public Map<String, KPI> getConfiguredKPIs(@NonNull Map<String, KPIConfig> kpiConfigMap) {
 
-        Map<String, KPIConfig> config = landscape.getConfig().getKPIs();
-        Map<String, KPI> kpis = new HashMap<>(defaultKPIs.size() + config.size());
+        Objects.requireNonNull(kpiConfigMap, "kpi config is null or empty");
```
Comment (mfbieber) [inline src/main/java/de/bonndan/nivio/assessment/kpi/KPIFactory.java:38]: `Objects.requireNonNull()` does not check if the map is empty, only if an object == null.

**Comment by mfbieber** [comment]
While reviewing I came across `reIndex(final LandscapeDescription input)` in `Indexer`. I tried to figure out if we also have the `index()` case with a `LandscapeDescription`, but did not find it.

Does that case exist?

And do the oder `index()` methods we have basically do the same thing as the `reIndex()`? Or asked differently: is "index" the same thing for all these methods?

**Comment by bonndan** [comment]
We had no "index" method until now. If chose the name to indicate that the method would also update existing landscape, but now there is a comment and the method is named "index". 👍

File: src/main/java/de/bonndan/nivio/assessment/kpi/KPIFactory.java (hunk 2 above)
Comment (bonndan) [reply src/main/java/de/bonndan/nivio/assessment/kpi/KPIFactory.java:38]: Yes. If the are no configured KPIs, the map would be empty.

Comment (mfbieber) [reply src/main/java/de/bonndan/nivio/assessment/kpi/KPIFactory.java:38]: I meant that the message is missleading, we are not checking if the map is empty, only if it is null ;)

Comment (bonndan) [reply src/main/java/de/bonndan/nivio/assessment/kpi/KPIFactory.java:38]: Fixed in a different branch.

**Pull request merged**